linera-core/src/client.rs

```diff
@@ -68,41 +68,47 @@ use crate::{
 mod client_tests;
 
 /// A builder that creates `ChainClients` which share the cache and notifiers.
-pub struct Client<ValidatorNodeProvider> {
+pub struct Client<ValidatorNodeProvider, Storage> {
     /// How to talk to the validators.
     validator_node_provider: ValidatorNodeProvider,
+    /// Local node to manage the execution state and the local storage of the chains that we are
+    /// tracking.
+    local_node: LocalNodeClient<Storage>,
     /// Maximum number of pending messages processed at a time in a block.
     max_pending_messages: usize,
     /// The policy for automatically handling incoming messages.
     message_policy: MessagePolicy,
     /// Whether to block on cross-chain message delivery.
     cross_chain_message_delivery: CrossChainMessageDelivery,
-    /// Cached hashed certificate values by hash.
-    recent_hashed_certificate_values: Arc<ValueCache<CryptoHash, HashedCertificateValue>>,
-    /// Cached blobs by blob ID.
-    recent_hashed_blobs: Arc<ValueCache<BlobId, HashedBlob>>,
-    /// One-shot channels to notify callers when messages of a particular chain have been
-    /// delivered.
-    delivery_notifiers: Arc<tokio::sync::Mutex<DeliveryNotifiers>>,
     /// References to clients waiting for chain notifications.
     notifier: Arc<Notifier<Notification>>,
 }
 
-impl<ValidatorNodeProvider: Clone> Client<ValidatorNodeProvider> {
+impl<P, S: Storage + Clone> Client<P, S> {
     /// Creates a new `Client` with a new cache and notifiers.
     pub fn new(
-        validator_node_provider: ValidatorNodeProvider,
+        validator_node_provider: P,
+        storage: S,
         max_pending_messages: usize,
         cross_chain_message_delivery: CrossChainMessageDelivery,
     ) -> Self {
+        let state = WorkerState::new_for_client(
+            "Client node".to_string(),
+            storage,
+            Arc::new(ValueCache::default()),
+            Arc::new(ValueCache::default()),
+            Arc::new(tokio::sync::Mutex::new(DeliveryNotifiers::default())),
```

Comment:
Do we still need these `Arc`s ?

Reply:
We don't, with some additional refactoring, but in the interests of keeping the PR minimal I'm planning on doing this in a later PR.

Reply:
In fact, if I remember correctly, these are now properly owned by the `LocalNodeClient`, and don't need to be here at all.

Reply:
Addressed in https://github.com/linera-io/linera-protocol/pull/2133.